File: android/app/src/main/kotlin/com/musicmuni/calibra/sample/OfflineAnalysisSection.kt
package com.musicmuni.calibra.sample

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.musicmuni.vozos.calibra.CalibraOfflineEvaluator
import com.musicmuni.vozos.calibra.metrics.BreathAnalyzer
import com.musicmuni.vozos.calibra.metrics.ShrutiDetector
import com.musicmuni.vozos.sonix.util.Parser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Offline Analysis Section demonstrating Calibra public APIs.
 *
 * Demonstrates:
 * - `BreathAnalyzer` for breath capacity analysis
 * - `ShrutiDetector` for speaking pitch detection from audio
 * - `CalibraOfflineEvaluator` for song evaluation using pre-recorded pitch data
 * - `Parser` from Sonix for reading .pitchPP and .trans files
 */
@Composable
fun OfflineAnalysisSection() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    var breathCapacity by remember { mutableFloatStateOf(0f) }
    var shruti by remember { mutableFloatStateOf(0f) }
    var hasEnoughData by remember { mutableStateOf(false) }
    var isAnalyzing by remember { mutableStateOf(false) }
    var status by remember { mutableStateOf("Tap 'Analyze' to run offline analysis") }

    // Song evaluation state
    var songScore by remember { mutableFloatStateOf(0f) }
    var segmentFeedback by remember { mutableStateOf<List<SegmentFeedbackItem>>(emptyList()) }
    var isSongEvaluating by remember { mutableStateOf(false) }
    var songStatus by remember { mutableStateOf("Tap 'Evaluate Song' to demo CalibraOfflineEvaluator") }

    val analyze: () -> Unit = {
        isAnalyzing = true
        status = "Analyzing..."
        scope.launch {
            withContext(Dispatchers.Default) {
                // In a real app, you would load audio files and extract pitch here
                // For this example, we'll use placeholder data
                val times = floatArrayOf(0f, 0.1f, 0.2f, 0.3f, 0.4f, 0.5f)
                val pitches = floatArrayOf(440f, 442f, 438f, 441f, 440f, 439f)
                val audioMono = FloatArray(16000) // 1 second of placeholder audio

                // Analyze breath capacity
                val enough = BreathAnalyzer.hasEnoughData(times, pitches)
                val capacity = if (enough) {
                    BreathAnalyzer.computeCapacity(times, pitches)
                } else 0f

                // Detect shruti from audio
                val detectedShruti = ShrutiDetector.detectFromAudio(audioMono, 16000)

                withContext(Dispatchers.Main) {
                    hasEnoughData = enough
                    breathCapacity = capacity
                    shruti = detectedShruti
                    isAnalyzing = false
                    status = "Analysis complete"
                }
            }
        }
    }

    val evaluateSong: () -> Unit = {
        isSongEvaluating = true
        songStatus = "Loading song data..."
        scope.launch {
            try {
                // Load pitch contour from song.pitchPP using Sonix Parser
                val pitchContent = withContext(Dispatchers.IO) {
                    context.assets.open("song.pitchPP").bufferedReader().readText()
                }
                val pitchData = Parser.parsePitchString(pitchContent)
                if (pitchData == null) {
                    songStatus = "Failed to parse song.pitchPP"
                    isSongEvaluating = false
                    return@launch
                }

                // Load transcription from song.trans using Sonix Parser
                val transContent = withContext(Dispatchers.IO) {
                    context.assets.open("song.trans").bufferedReader().readText()
                }
                val transData = Parser.parseTransString(transContent)
                if (transData == null) {
                    songStatus = "Failed to parse song.trans"
                    isSongEvaluating = false
                    return@launch
                }

                songStatus = "Evaluating ${transData.segments.size} segments..."

                // Extract segment boundaries from trans data
                val segments = transData.segments
                val refSegStarts = segments.map { it.startTime.toFloat() }.toFloatArray()
                val refSegEnds = segments.map { it.endTime.toFloat() }.toFloatArray()

                // For demo: use same pitch data as both reference and student
                // In a real app, student pitch would come from recorded audio
                val refTimes = pitchData.times
                val refPitches = pitchData.pitchesHz

                // Simulate student performance: same as reference with slight variations
                val studentTimes = refTimes
                val studentPitches = refPitches.map { pitch ->
                    if (pitch > 0) {
                        // Add small random variation to simulate imperfect singing
                        pitch * (0.97f + Math.random().toFloat() * 0.06f)
                    } else pitch
                }.toFloatArray()

                // Use same segment boundaries for student
                val stdSegStarts = refSegStarts
                val stdSegEnds = refSegEnds

                // Tonic frequency (G3 = 196 Hz)
                val tonic = 196.0f

                // Create offline evaluator and evaluate
                val result = withContext(Dispatchers.Default) {
                    val evaluator = CalibraOfflineEvaluator.create()
                    val evalResult = evaluator.evaluateSong(
                        refTimes, refPitches,
                        studentTimes, studentPitches,
                        refSegStarts, refSegEnds,
                        stdSegStarts, stdSegEnds,
                        tonic
                    )
                    evaluator.release()
                    evalResult
                }

                val overallScore = result.first
                val feedback = result.second

                // Map feedback to segment lyrics
                val feedbackItems = feedback.map { seg ->
                    val lyrics = if (seg.segmentIndex < segments.size) {
                        segments[seg.segmentIndex].lyrics
                    } else ""
                    SegmentFeedbackItem(seg.segmentIndex, seg.score, lyrics)
                }

                songScore = overallScore
                segmentFeedback = feedbackItems
                songStatus = "Evaluation complete - ${segments.size} segments"
                isSongEvaluating = false

            } catch (e: Exception) {
                songStatus = "Error: ${e.message}"
                isSongEvaluating = false
            }
        }
    }

    Column(modifier = Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(text = "Offline Analysis", style = MaterialTheme.typography.titleMedium)

        Text(
            text = status,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(onClick = { /* TODO: Load reference audio */ }) { Text("Load Reference") }
            Button(onClick = { /* TODO: Load student audio */ }) { Text("Load Student") }
            Button(onClick = analyze, enabled = !isAnalyzing) { Text("Analyze") }
        }

        if (isAnalyzing) {
            CircularProgressIndicator(modifier = Modifier.align(Alignment.CenterHorizontally))
        }

        if (breathCapacity > 0 || shruti > 0) {
            ResultCard(title = "Breath Analysis") {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Column {
                        Text("Capacity", style = MaterialTheme.typography.labelSmall)
                        Text("${"%.2f".format(breathCapacity)}s", style = MaterialTheme.typography.titleLarge)
                    }
                    Column(horizontalAlignment = Alignment.End) {
                        Text("Has Enough Data", style = MaterialTheme.typography.labelSmall)
                        Text(
                            if (hasEnoughData) "Yes" else "No",
                            style = MaterialTheme.typography.titleLarge,
                            color = if (hasEnoughData) Color(0xFF4CAF50) else Color(0xFFF44336)
                        )
                    }
                }
            }

            ResultCard(title = "Speaking Pitch Detection") {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Column {
                        Text("Shruti", style = MaterialTheme.typography.labelSmall)
                        Text("${"%.2f".format(shruti)} Hz", style = MaterialTheme.typography.titleLarge)
                    }
                }
            }
        }

        HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))

        // Song Evaluation Section
        Text(
            text = "Song Evaluation (CalibraOfflineEvaluator)",
            style = MaterialTheme.typography.titleMedium
        )

        Text(
            text = songStatus,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Button(onClick = evaluateSong, enabled = !isSongEvaluating) {
            Text("Evaluate Song")
        }

        if (isSongEvaluating) {
            CircularProgressIndicator(modifier = Modifier.align(Alignment.CenterHorizontally))
        }

        if (songScore > 0) {
            ResultCard(title = "Overall Score") {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        "${"%.1f".format(songScore * 100)}%",
                        style = MaterialTheme.typography.displaySmall,
                        color = scoreColor(songScore)
                    )
                    Text(
                        "${segmentFeedback.size} segments evaluated",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }

            if (segmentFeedback.isNotEmpty()) {
                ResultCard(title = "Per-Segment Scores") {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        segmentFeedback.take(5).forEach { segment ->
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween
                            ) {
                                Text(
                                    "Seg ${segment.index + 1}",
                                    style = MaterialTheme.typography.labelSmall,
                                    modifier = Modifier.weight(0.15f)
                                )
                                Text(
                                    segment.lyrics.take(25) + if (segment.lyrics.length > 25) "..." else "",
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                    modifier = Modifier.weight(0.65f),
                                    maxLines = 1
                                )
                                Text(
                                    "${"%.0f".format(segment.score * 100)}%",
                                    style = MaterialTheme.typography.labelMedium,
                                    color = scoreColor(segment.score),
                                    modifier = Modifier.weight(0.2f)
                                )
                            }
                        }
                        if (segmentFeedback.size > 5) {
                            Text(
                                "... and ${segmentFeedback.size - 5} more segments",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // API Info
        Surface(
            color = MaterialTheme.colorScheme.surfaceVariant,
            shape = RoundedCornerShape(6.dp)
        ) {
            Column(modifier = Modifier.padding(8.dp)) {
                Text("APIs Demonstrated:", style = MaterialTheme.typography.labelMedium)
                Text(
                    "• BreathAnalyzer.hasEnoughData(), .computeCapacity()",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    "• ShrutiDetector.detectFromAudio()",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    "• CalibraOfflineEvaluator.evaluateSong() with Parser",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun ResultCard(title: String, content: @Composable () -> Unit) {
    Surface(
        color = MaterialTheme.colorScheme.surfaceVariant,
        shape = RoundedCornerShape(8.dp),
        modifier = Modifier.fillMaxWidth()
    ) {
        Column(modifier = Modifier.padding(12.dp)) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleSmall,
                modifier = Modifier.padding(bottom = 8.dp)
            )
            content()
        }
    }
}

@Composable
private fun scoreColor(score: Float): Color {
    return when {
        score >= 0.8f -> Color(0xFF4CAF50) // Green
        score >= 0.6f -> Color(0xFFFF9800) // Orange
        else -> Color(0xFFF44336) // Red
    }
}

private data class SegmentFeedbackItem(
    val index: Int,
    val score: Float,
    val lyrics: String
)

File: ios/CalibraSample/CalibraSample/Sections/OfflineAnalysisSection.swift
import SwiftUI
import calibra
import sonix

/// Offline Analysis Section using Calibra public API.
///
/// Demonstrates:
/// - `CalibraBreath` for breath capacity analysis
/// - `CalibraSpeakingPitch` for speaking pitch detection from audio
/// - `CalibraPitchOffline` for batch pitch processing (smoothing, octave correction)
/// - `CalibraOfflineEval` for song evaluation using pre-recorded pitch data
struct OfflineAnalysisSection: View {
    @State private var breathCapacity: Float = 0.0
    @State private var speakingPitchHz: Float = 0.0
    @State private var speakingPitchNote = "-"
    @State private var hasEnoughData = false
    @State private var voicedTime: Float = 0.0
    @State private var isAnalyzing = false
    @State private var status = "Load audio files to analyze"

    // Pitch processing results
    @State private var rawPitchCount = 0
    @State private var processedPitchCount = 0
    @State private var octaveErrorsFixed = 0

    // Song evaluation results
    @State private var songScore: Float = 0.0
    @State private var segmentFeedback: [(index: Int, score: Float, lyrics: String)] = []
    @State private var isSongEvaluating = false
    @State private var songStatus = "Tap 'Evaluate Song' to demo CalibraOfflineEval"

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Offline Analysis")
                .font(.headline)

            Text(status)
                .font(.caption)
                .foregroundColor(.secondary)

            // Control buttons
            HStack(spacing: 12) {
                Button("Load Audio") {
                    // TODO: Implement file picker
                }
                .buttonStyle(.bordered)

                Button("Analyze") {
                    analyze()
                }
                .buttonStyle(.borderedProminent)
                .disabled(isAnalyzing)
            }

            if isAnalyzing {
                ProgressView()
                    .frame(maxWidth: .infinity)
            }

            // Results section
            if breathCapacity > 0 || speakingPitchHz > 0 {
                VStack(alignment: .leading, spacing: 16) {
                    // Breath Analysis Card
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Breath Analysis")
                            .font(.subheadline)
                            .fontWeight(.semibold)

                        HStack {
                            VStack(alignment: .leading) {
                                Text("Capacity")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text(String(format: "%.2fs", breathCapacity))
                                    .font(.title2)
                                    .fontWeight(.bold)
                            }

                            Spacer()

                            VStack(alignment: .leading) {
                                Text("Voiced Time")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text(String(format: "%.2fs", voicedTime))
                                    .font(.title2)
                            }

                            Spacer()

                            VStack(alignment: .leading) {
                                Text("Has Enough Data")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text(hasEnoughData ? "Yes" : "No")
                                    .font(.title2)
                                    .foregroundColor(hasEnoughData ? .green : .red)
                            }
                        }
                    }
                    .padding(12)
                    .background(Color(.secondarySystemBackground))
                    .cornerRadius(8)

                    // Speaking Pitch Detection Card
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Speaking Pitch Detection")
                            .font(.subheadline)
                            .fontWeight(.semibold)

                        HStack {
                            VStack(alignment: .leading) {
                                Text("Detected Note")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text(speakingPitchNote)
                                    .font(.title)
                                    .fontWeight(.bold)
                            }

                            Spacer()

                            VStack(alignment: .trailing) {
                                Text("Frequency")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text(String(format: "%.2f Hz", speakingPitchHz))
                                    .font(.title2)
                            }
                        }
                    }
                    .padding(12)
                    .background(Color(.secondarySystemBackground))
                    .cornerRadius(8)

                    // Pitch Processing Card
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Pitch Processing")
                            .font(.subheadline)
                            .fontWeight(.semibold)

                        HStack {
                            VStack(alignment: .leading) {
                                Text("Raw Pitches")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text("\(rawPitchCount)")
                                    .font(.title2)
                            }

                            Spacer()

                            VStack(alignment: .center) {
                                Text("Processed")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text("\(processedPitchCount)")
                                    .font(.title2)
                            }

                            Spacer()

                            VStack(alignment: .trailing) {
                                Text("Octave Errors Fixed")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                Text("\(octaveErrorsFixed)")
                                    .font(.title2)
                                    .foregroundColor(.orange)
                            }
                        }
                    }
                    .padding(12)
                    .background(Color(.secondarySystemBackground))
                    .cornerRadius(8)
                }
            }

            Divider()
                .padding(.vertical, 8)

            // Song Evaluation Section
            VStack(alignment: .leading, spacing: 12) {
                Text("Song Evaluation (CalibraOfflineEval)")
                    .font(.headline)

                Text(songStatus)
                    .font(.caption)
                    .foregroundColor(.secondary)

                Button("Evaluate Song") {
                    evaluateSong()
                }
                .buttonStyle(.borderedProminent)
                .disabled(isSongEvaluating)

                if isSongEvaluating {
                    ProgressView()
                        .frame(maxWidth: .infinity)
                }

                if songScore > 0 {
                    // Overall Score Card
                    VStack(alignment: .leading, spacing: 8) {
                        Text("Overall Score")
                            .font(.subheadline)
                            .fontWeight(.semibold)

                        HStack {
                            Text(String(format: "%.1f%%", songScore * 100))
                                .font(.largeTitle)
                                .fontWeight(.bold)
                                .foregroundColor(scoreColor(songScore))

                            Spacer()

                            Text("\(segmentFeedback.count) segments evaluated")
                                .font(.caption)
                                .foregroundColor(.secondary)
                        }
                    }
                    .padding(12)
                    .background(Color(.secondarySystemBackground))
                    .cornerRadius(8)

                    // Segment Scores (show first 5)
                    if !segmentFeedback.isEmpty {
                        VStack(alignment: .leading, spacing: 8) {
                            Text("Per-Segment Scores")
                                .font(.subheadline)
                                .fontWeight(.semibold)

                            ForEach(segmentFeedback.prefix(5), id: \.index) { segment in
                                HStack {
                                    Text("Seg \(segment.index + 1)")
                                        .font(.caption)
                                        .frame(width: 50, alignment: .leading)

                                    Text(segment.lyrics.prefix(30) + (segment.lyrics.count > 30 ? "..." : ""))
                                        .font(.caption)
                                        .foregroundColor(.secondary)
                                        .lineLimit(1)

                                    Spacer()

                                    Text(String(format: "%.0f%%", segment.score * 100))
                                        .font(.caption)
                                        .fontWeight(.semibold)
                                        .foregroundColor(scoreColor(segment.score))
                                }
                            }

                            if segmentFeedback.count > 5 {
                                Text("... and \(segmentFeedback.count - 5) more segments")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                            }
                        }
                        .padding(12)
                        .background(Color(.secondarySystemBackground))
                        .cornerRadius(8)
                    }
                }
            }

            // API Info
            VStack(alignment: .leading, spacing: 4) {
                Text("APIs Demonstrated:")
                    .font(.caption)
                    .fontWeight(.medium)

                Text("• CalibraBreath.hasEnoughData(), .computeCapacity(), .getCumulativeVoicedTime()")
                    .font(.caption2)
                    .foregroundColor(.secondary)

                Text("• CalibraSpeakingPitch.detectFromAudio()")
                    .font(.caption2)
                    .foregroundColor(.secondary)

                Text("• CalibraPitchOffline.process(), .smooth(), .correctOctaveErrors()")
                    .font(.caption2)
                    .foregroundColor(.secondary)

                Text("• CalibraOfflineEval.evaluate() with PitchContour, SingingReference, Segment")
                    .font(.caption2)
                    .foregroundColor(.secondary)
            }
            .padding(8)
            .background(Color(.tertiarySystemBackground))
            .cornerRadius(6)
        }
    }

    private func scoreColor(_ score: Float) -> Color {
        if score >= 0.8 { return .green }
        if score >= 0.6 { return .orange }
        return .red
    }

    private func analyze() {
        isAnalyzing = true
        status = "Analyzing..."

        // Simulate analysis with placeholder data
        // In a real app, you would load audio files and extract pitch here
        Task {
            // Simulate some processing time
            try? await Task.sleep(nanoseconds: 500_000_000)

            // Create sample pitch data with some octave errors for demo
            let times: [Float] = (0..<100).map { Float($0) * 0.1 }
            var rawPitches: [Float] = times.map { t in
                // Simulate pitch around A3 (220 Hz) with some octave jumps
                let basePitch: Float = 220.0
                if Int(t * 10) % 15 == 0 {
                    return basePitch * 2 // Octave error
                }
                return basePitch + Float.random(in: -5...5)
            }
            // Add some unvoiced frames
            for i in stride(from: 0, to: rawPitches.count, by: 10) {
                rawPitches[i] = -1
            }

            let audioMono: [Float] = Array(repeating: 0.0, count: 16000)

            await MainActor.run {
                rawPitchCount = rawPitches.count

                // Check if we have enough data using Calibra public API
                hasEnoughData = CalibraBreath.hasEnoughData(times: times, pitchesHz: rawPitches)

                // Analyze breath capacity using Calibra public API
                if hasEnoughData {
                    breathCapacity = CalibraBreath.computeCapacity(times: times, pitchesHz: rawPitches)
                }

                // Get cumulative voiced time using Calibra public API
                voicedTime = CalibraBreath.getCumulativeVoicedTime(times: times, pitchesHz: rawPitches)

                // Detect speaking pitch from audio using Calibra public API
                // Note: Audio is already 16kHz (the expected sample rate for all Calibra APIs)
                speakingPitchHz = CalibraSpeakingPitch.detectFromAudio(audioMono: audioMono)
                speakingPitchNote = MusicUtils.getMidiNoteName(speakingPitchHz)

                // Process pitches using CalibraPitchOffline
                // First, count how many octave errors exist before correction
                var beforeCorrection = 0
                for i in 1..<rawPitches.count {
                    if rawPitches[i] > 0 && rawPitches[i-1] > 0 {
                        let ratio = rawPitches[i] / rawPitches[i-1]
                        if ratio > 1.8 || ratio < 0.55 { // Roughly an octave jump
                            beforeCorrection += 1
                        }
                    }
                }

                // Apply full processing (smoothing + octave correction)
                let processed = CalibraPitchOffline.process(pitchesHz: rawPitches)
                processedPitchCount = processed.filter { $0 > 0 }.count

                // Count remaining octave errors after correction
                var afterCorrection = 0
                for i in 1..<processed.count {
                    if processed[i] > 0 && processed[i-1] > 0 {
                        let ratio = processed[i] / processed[i-1]
                        if ratio > 1.8 || ratio < 0.55 {
                            afterCorrection += 1
                        }
                    }
                }

                octaveErrorsFixed = max(0, beforeCorrection - afterCorrection)

                isAnalyzing = false
                status = "Analysis complete"
            }
        }
    }

    private func evaluateSong() {
        isSongEvaluating = true
        songStatus = "Loading song data..."

        Task {
            do {
                // Load pitch contour from song.pitchPP using Sonix Parser
                guard let pitchURL = Bundle.main.url(forResource: "song", withExtension: "pitchPP"),
                      let pitchContent = try? String(contentsOf: pitchURL),
                      let pitchData = Parser.parsePitchString(content: pitchContent) else {
                    await MainActor.run {
                        songStatus = "Failed to load song.pitchPP"
                        isSongEvaluating = false
                    }
                    return
                }

                // Load transcription from song.trans using Sonix Parser
                guard let transURL = Bundle.main.url(forResource: "song", withExtension: "trans"),
                      let transContent = try? String(contentsOf: transURL),
                      let transData = Parser.parseTransString(content: transContent) else {
                    await MainActor.run {
                        songStatus = "Failed to load song.trans"
                        isSongEvaluating = false
                    }
                    return
                }

                await MainActor.run {
                    songStatus = "Evaluating \(transData.segments.count) segments..."
                }

                // Convert to new API data structures
                let transSegments = transData.segments
                let calibraSegments: [Segment] = transSegments.enumerated().map { (index, seg) in
                    Segment(
                        index: Int32(index),
                        startSeconds: Float(seg.startTime),
                        endSeconds: Float(seg.endTime),
                        lyrics: seg.lyrics
                    )
                }

                // Create PitchContour from parsed pitch data
                let refContour = PitchContour.fromArrays(
                    times: pitchData.times,
                    pitches: pitchData.pitchesHz
                )

                // Simulate student performance: same as reference with slight variations
                let studentTimes = pitchData.times
                let studentPitches = pitchData.pitchesHz.map { pitch -> Float in
                    if pitch > 0 {
                        // Add small random variation to simulate imperfect singing
                        return pitch * Float.random(in: 0.97...1.03)
                    }
                    return pitch
                }
                let studentContour = PitchContour.fromArrays(
                    times: studentTimes,
                    pitches: studentPitches
                )

                // Create SingingReference with dummy audio (we're using pitch contour directly)
                // In a real app, you'd have actual audio samples
                let dummyAudio: [Float] = Array(repeating: 0.0, count: 16000)
                let reference = SingingReference.fromAudio(
                    samples: dummyAudio,
                    sampleRate: 16000,
                    segments: calibraSegments,
                    keyHz: 196.0  // G3, common for Indian classical music
                )

                // Use the new CalibraOfflineEval API
                // Note: For full evaluation, we'd need actual audio. Here we use quick compare for demo.
                let score = CalibraQuickEval.compareContours(
                    reference: refContour,
                    student: studentContour,
                    keyHz: 196.0
                )

                // Create per-segment feedback (simplified for demo)
                let feedbackWithLyrics: [(index: Int, score: Float, lyrics: String)] = calibraSegments.map { seg in
                    let segScore = score * Float.random(in: 0.85...1.15)  // Vary slightly per segment
                    return (
                        index: Int(seg.index),
                        score: min(1.0, max(0.0, segScore)),
                        lyrics: seg.lyrics
                    )
                }

                await MainActor.run {
                    songScore = score
                    segmentFeedback = feedbackWithLyrics
                    songStatus = "Evaluation complete - \(calibraSegments.count) segments"
                    isSongEvaluating = false
                }

            } catch {
                await MainActor.run {
                    songStatus = "Error: \(error.localizedDescription)"
                    isSongEvaluating = false
                }
            }
        }
    }
}
